### DISCUSSION TITLE
Why do you use url-connection-client instead of apache-client as SdkHttpClient?

### DISCUSSION THREAD

**Discussion by schabe77** [aws-powertools/powertools-lambda-java#1093] [Q&A] [answered]
Hi,

I wanted to use your powertools-cloudformation and the first thing that happened was that an exception was thrown:

    software.amazon.awssdk.core.exception.SdkClientException: Multiple HTTP implementations were found on the classpath. To avoid non-deterministic loading implementations, please explicitly provide an HTTP client via the client builders set the software.amazon.awssdk.http.service.impl system property with the FQCN of the HTTP service to use as the default, or remove all but one HTTP implementation from the classpath.

The reason was that beside software.amazon.awssdk:url-connection-client (that comes as dependency of powertools-cloudformation) I also have software.amazon.awssdk:apache-client in my classpath (that comes with software.amazon.awssdk:secretsmanager and software.amazon.awssdk:rds). I circumvented this issue by setting the system property as suggested, but I wondered if would be possible for you to switch the HttpClient.

I don't know if apache-client SdkHttpClient is the standard client of awssdk, I just wanted to know if there is a specific reason why you use url-connection-client while rds and secretsmanager use software.amazon.awssdk:apache-client.

**Comment by msailes** [comment, marked as answer]
Hi @schabe77,

Thanks for the feedback, let me try and explain our thinking.

As you've seen there are multiple HTTP clients available to use with the SDK. Lambda Powertools for Java tries to help by picking the best option when multiple exist, while still allowing customisation. 

The url-connection-client meets our [tenet](https://awslabs.github.io/aws-lambda-powertools-java/#tenets) to keep our technology choices lean. It is the lowest memory usage client and has the fastest start up time, often a very important characteristic when using Lambda.

From your feedback it looks like we've still got some work to do with meeting another of our tenets (Eases the adoption of best practices). If you could raise an issue with the problem you had that would be really helpful.

Further on this subject, with the recent GA of the CRT HTTP client https://github.com/awslabs/aws-lambda-powertools-java/issues/1092 we will investigate whether this is a further improvement.

Thanks,

Mark

**Reply by schabe77** [reply]
Hi Mark,

thank you for the quick reply. I have no problems with the solution. I was just wondering if the different client was used intentionally or by mistake. It just bothered me because I had to manually set a system property, which I always try to avoid. But if it is well thought out and has its reasons, then I have no problem with it.

Thanks,

Christian
